packages/core/src/Renderer.ts

```diff
@@ -79,6 +97,146 @@ export abstract class Renderer extends Component {
+   * @returns Material
+   */
+  getMaterial(index: number): Material | null;
+
+  getMaterial(index: number = 0): Material | null {
+    const materials = this._materials;
+    return materials.length > 0 ? materials[index] : null;
+  }
+
+  /**
+   * Set the first material.
+   * @param material - The first material
+   */
+  setMaterial(material: Material);
+
+  /**
+   * Set material by index.
+   * @param index - Material index
+   * @param material - The material
+   */
+  setMaterial(index: number, material: Material);
+
+  setMaterial(indexOrMaterial: number | Material, material: Material = null) {
+    let index;
+    if (typeof indexOrMaterial === "number") {
+      index = indexOrMaterial;
+    } else {
+      index = 0;
+      material = indexOrMaterial;
+    }
+
+    const materials = this._materials;
+    if (index >= materials.length) {
+      materials.length = index + 1;
+    }
+
+    const materialsInstance = this._materialsInstanced;
+    const internalMaterial = materials[index];
+    if (internalMaterial !== material) {
+      materials[index] = material;
+      index < materialsInstance.length && (materialsInstance[index] = false);
+      internalMaterial && internalMaterial._addRefCount(-1);
+      material && material._addRefCount(1);
+    }
+  }
+
+  /**
+   * Get all instance materials.
+   * @remarks Calling this function for the first time after the material is set will create an instance material to ensure that it is unique to the renderer.
+   * @returns All instance materials
+   */
+  getInstanceMaterials(): Readonly<Material[]> {
+    const materials = this._materials;
+    const materialsInstance = this._materialsInstanced;
+    for (let i = 0, n = materials.length; i < n; i++) {
+      if (!materialsInstance[i]) {
+        this._createInstanceMaterial(this._materials[i], i);
+      }
+    }
+    return materials;
+  }
+
+  /**
+   * Get all materials.
+   * @returns All materials
+   */
+  getMaterials(): Readonly<Material[]> {
+    return this._materials;
+  }
+
+  /**
+   * Set all materials.
+   * @param materials - All materials
+   */
+  setMaterials(materials: Material[]): void {
+    const count = materials.length;
+    const internalMaterials = this._materials;
+    const materialsInstanced = this._materialsInstanced;
+
+    for (let i = count, n = internalMaterials.length; i < n; i++) {
+      const internalMaterial = internalMaterials[i];
+      internalMaterial && internalMaterial._addRefCount(-1);
+    }
+
+    internalMaterials.length !== count && (internalMaterials.length = count);
+    materialsInstanced.length !== 0 && (materialsInstanced.length = 0);
+
+    for (let i = 0; i < count; i++) {
+      const internalMaterial = internalMaterials[i];
+      const material = materials[i];
+      if (internalMaterial !== material) {
+        internalMaterials[i] = material;
+        internalMaterial && internalMaterial._addRefCount(-1);
+        material && material._addRefCount(1);
+      }
```

Comment:
Can this be replaced with `this.setMaterial(i,material)`?

Reply:
Can! But some customized optimization must be lost.